docs/zkapps/tutorials/09-recursion.mdx

```diff
@@ -56,86 +56,310 @@ You build recursive zkApps with [ZkProgram](/zkapps/o1js-reference/modules/Exper
+- The first argument of a ZkProgram method is always the state of the ZkProgram, named `publicInput` since it is public.
+
+```typescript
+3 export const Add = ZkProgram({
+4  name: 'add-example',
+5  publicInput: Field,
+6
+7  methods: {
+8    init: {
+9      privateInputs: [],
+10
+11      method(state: Field) {
+12        state.assertEquals(Field(0));
+13      },
+14    },
+15  },
+16 });
+17
 ```
 
 Add another method that takes an existing proof, adds a new number to it, and produces a new proof:
 
 ```typescript
-    addNumber: {
-      privateInputs: [SelfProof, Field ],
-
-      method(newState: Field, earlierProof: SelfProof<Field>, numberToAdd: Field) {
-        earlierProof.verify();
-        newState.assertEquals(earlierProof.publicInput.add(numberToAdd));
-      },
-    },
+15
+16    addNumber: {
+17      privateInputs: [SelfProof, Field],
+18
+19      method(
+20        newState: Field,
+21        earlierProof: SelfProof<Field, void>,
+22        numberToAdd: Field
+23      ) {
+24        earlierProof.verify();
+25        newState.assertEquals(earlierProof.publicInput.add(numberToAdd));
+26      },
+27    },
+28  },
+29 });
 ```
 
 Use recursion to combine two proofs:
 
 ```typescript
-  add: {
-    privateInputs: [ SelfProof, SelfProof ],
-
-    method(
-      newState: Field,
-      earlierProof1: SelfProof<Field>,
-      earlierProof2: SelfProof<Field>,
-    ) {
-      earlierProof1.verify();
-      earlierProof2.verify();    	
-      newState.assertEquals(earlierProof1.publicInput.add(earlierProof2.publicInput));
-    },
-  },
+28
+29    add: {
+30      privateInputs: [SelfProof, SelfProof],
+31
+32      method(
+33        newState: Field,
+34        earlierProof1: SelfProof<Field, void>,
+35        earlierProof2: SelfProof<Field, void>
+36      ) {
+37        earlierProof1.verify();
+38        earlierProof2.verify();
+39        newState.assertEquals(
+40          earlierProof1.publicInput.add(earlierProof2.publicInput)
+41        );
+42      },
+43    },
+44  },
+45 });
 ```
 
-To use ZkProgram, compile it and then call methods on it:
+## Interact with a smart contract
+
+Next, write a script that interacts with your smart contract. As before, the complete [main.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/01-hello-world/src/main.ts) example file is provided. Follow these steps to build the `main.ts` file so you can interact with the smart contract.
+
+### Imports
+
+For this tutorial, the `import` statement brings in items from `o1js` that you use to interact with your smart contract. 
+
+1. Copy the following lines from [main.ts](https://github.com/o1-labs/docs2/blob/main/examples/zkapps/01-hello-world/src/main.ts) example file into the `src/main.ts` file:
+
+```ts src/main.ts
+1 import { Add } from './Add.js';
+2 import { Field, verify } from 'o1js';
+```
+
+### Call Methods
+
+Now we will add methods to call ZkProgram:
 
 ```typescript
-  console.log('compiling...');
+3
```

Comment:
be sure the example code in https://github.com/o1-labs/docs2/blob/main/examples/zkapps/09-recursion/src/main.ts is current 
and match here (omit line numbers please)